sycl/source/handler.cpp

```diff
@@ -27,8 +28,124 @@ namespace sycl {
+      detail::ExtendedMembersType::HANDLER_IMPL,
+      std::make_shared<detail::handler_impl>()};
+  ExtendedMembers->push_back(std::move(HandlerImplMember));
+  MSharedPtrStorage.push_back(std::move(ExtendedMembers));
+}
+
+/// Gets the handler_impl at the start of the extended members.
+// detail::GlobalHandler::instance().getHandlerExtendedMembersMutex() must
+// be held when calling this function.
+std::shared_ptr<detail::handler_impl>
+getHandlerImpl(const std::shared_ptr<std::vector<detail::ExtendedMemberT>>
+                   &ExtendedMembersVec) {
+  assert(ExtendedMembersVec->size() > 0);
+
+  auto HandlerImplMember = (*ExtendedMembersVec)[0];
+
+  assert(detail::ExtendedMembersType::HANDLER_IMPL == HandlerImplMember.MType);
+
+  return std::static_pointer_cast<detail::handler_impl>(
+      HandlerImplMember.MData);
+}
+
+// Common implementation for getting/inserting handler kernel bundle.
+// detail::GlobalHandler::instance().getHandlerExtendedMembersMutex() must
+// be held when calling this function.
+std::shared_ptr<detail::kernel_bundle_impl>
+getOrInsertHandlerKernelBundleCommon(
+    const std::shared_ptr<std::vector<detail::ExtendedMemberT>>
+        &ExtendedMembersVec,
+    const std::shared_ptr<detail::queue_impl> &Queue, bool Insert) {
+  // Look for the kernel bundle in extended members
+  std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImpPtr;
+  for (const detail::ExtendedMemberT &EMember : *ExtendedMembersVec)
+    if (detail::ExtendedMembersType::HANDLER_KERNEL_BUNDLE == EMember.MType) {
+      KernelBundleImpPtr =
+          std::static_pointer_cast<detail::kernel_bundle_impl>(EMember.MData);
+      break;
+    }
+
+  // No kernel bundle yet, create one
+  if (!KernelBundleImpPtr && Insert) {
+    KernelBundleImpPtr = detail::getSyclObjImpl(
+        get_kernel_bundle<bundle_state::input>(Queue->get_context()));
+    if (KernelBundleImpPtr->empty()) {
+      KernelBundleImpPtr = detail::getSyclObjImpl(
+          get_kernel_bundle<bundle_state::executable>(Queue->get_context()));
+    }
+
+    detail::ExtendedMemberT EMember = {
+        detail::ExtendedMembersType::HANDLER_KERNEL_BUNDLE, KernelBundleImpPtr};
+    ExtendedMembersVec->push_back(EMember);
+  }
+
+  return KernelBundleImpPtr;
+}
+
+// If the submission state is SPEC_CONST_SET_STATE this function returns false.
+// Otherwise it sets the submission state to EXPLICIT_KERNEL_BUNDLE_STATE and
+// returns true.
+bool handler::setStateExplicitKernel() {
+  std::lock_guard<std::mutex> Lock(
+      detail::GlobalHandler::instance().getHandlerExtendedMembersMutex());
+
+  assert(!MSharedPtrStorage.empty());
+
+  std::shared_ptr<std::vector<detail::ExtendedMemberT>> ExendedMembersVec =
+      detail::convertToExtendedMembers(MSharedPtrStorage[0]);
+
+  auto HandlerImpl = getHandlerImpl(ExendedMembersVec);
+  if (HandlerImpl->MSubmissionState ==
+      detail::HandlerSubmissionState::SPEC_CONST_SET_STATE)
+    return false;
```

Comment:
Can we throw an exception right here?

Reply:
Logic has been moved to `handler_impl` and the exceptions for setting state has been moved to there. The exception in `get_specialization_constant` stays in handler.hpp.